### DISCUSSION TITLE
Can anyone provide a summary on how code completion works with ALE?

### DISCUSSION THREAD

**Discussion by gnietto** [w0rp/ale#4682] [Q&A - Ask for help with problems]
@hsanson @w0rp and all the people:
I am user of Neovim or Vim editors and code completion feature doesn't always work.
I know code completion depends on LSP that supports that feature.
As far as I know, rust-analyzer supports code completion (doesn't work), it appears pyls also supports it, and tsserver appears to work slightly right.
I've read documentation about code completion with ALE, but I feel confused reading alternatives to improve the feature:
  . Automatic completion support with ALE
  . Omnicompletion c-x c-o
  . Integration with asyncomplete
  . Integration with deoplete
Can anyone provide a summary or an explicit workflow explanation on how code completion works or should work using ALE?
Thanks in advance.

**Comment by gcb** [comment]
seems very complete https://github.com/dense-analysis/ale?tab=readme-ov-file#completion

either [Deoplete](https://github.com/Shougo/deoplete.nvim) or `let g:ale_completion_enabled = 1; set omnifunc=ale#completion#OmniFunc` and then `<C-x><C-o>`

Note that you must have a linter enabled otherwise C-xC-o will just enter Insert mode and do nothing.

for rust you must add `let b:ale_linters = ['analyzer']`... still learning how to enable linters automatically, for now i just have a autocmd for .rs files.
